Correct trigger param parsing

SCdF opened the pull request:
If `params` doesn't parse right we were just dropping the error (yay callbacks!). I've refactored the code so that the error can be passed back through the main callback correctly.

First commit is just the tests, second is the code what fixes it.

![image](https://cloud.githubusercontent.com/assets/583851/24370568/c21f0cd8-131f-11e7-96ac-ba53a7cb07f0.png)

garethbowen commented on transitions/registration.js:
```diff
@@ -218,36 +218,44 @@ module.exports = {
+                            try {
+                                options.params = event.params.split(',');
+                            } catch (e) {
+                                return cb(new Error('event params cannot be parsed: must be either a comma-delimited String, or JSON *inside* a String'));
```
"JSON inside a string" doesn't make sense to me. Can you say "serialized JSON" or is that too technical?

garethbowen replied:
Also, maybe include both `e`s in the error somewhere - they might be useful.

SCdF replied:
Yeah that was the least technical I could think to make that error on one line. However, whatever, let's make it really detailed with multiple lines and examples! It's the multi-line-string future!

SCdF commented:
@garethbowen take another look please. I changed my approach a reasonable amount. The previous approach wouldn't actually detect *bad* JSON because it would just fall back to splitting a string.

So if you passed: `"{\"foo\": \"bar\""` (missing the closing `}`) it wouldn't actually error, because it would `JSON.parse()` that, fail, then `.split(',')` it, which would result in params equalling the entire string. Since params aren't required necessarily it would just happy process the doc and move on with its life.

This attempt does two things differently:
 - It supports mapping raw JSON. I'm not convinced our entire app will support this properly due to kanso.json expecting a string here, so it's still not our official method, but it's there anyway
 - We try to detect if your string is supposed to be JSON and just parse it that way, instead of falling back on string.split
